modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -65,6 +71,11 @@ ComputeFrictionalForceLMMechanicalContact::ComputeFrictionalForceLMMechanicalCon
     if (_friction_var->feType().order != static_cast<Order>(0))
       mooseError(
           "Frictional contact constraints only support elemental variables of CONSTANT order");
+
+  if (_3d && !_friction_var_dir)
+    paramError("friction_lm_dir",
+               "Three-dimensional mortar frictional contact simulations require an additional "
+               "frictional Lagrange's multiplier to enforce a second tangential pressure");
```

Comment:
Cool to see this getting fixed up! I'm excited to see this, I'll take a closer look when you're ready, just wanted to point out that below I neglected to compute the z component of the relative velocity, that should probably be added to the `if (_has_disp_z)` part below where the derivative is being calculated.

Reply:
It should be fixed now.